From: Josselin Mouette <joss@debian.org>
Date: Mon, 16 Aug 2010 18:14:38 +0200
Subject: Avoid spamming the console for using deprecated functions

---
 gobject/pygobject.c | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/gobject/pygobject.c b/gobject/pygobject.c
index 70dc89a..456f743 100644
--- a/gobject/pygobject.c
+++ b/gobject/pygobject.c
@@ -182,9 +182,6 @@ pygobject_register_sinkfunc(GType type, void (* sinkfunc)(GObject *object))
 {
     SinkFunc sf;
 
-    g_message ("pygobject_register_sinkfunc is deprecated (%s)",
-               g_type_name(type));
-
 #if 0
     g_return_if_fail(G_TYPE_IS_OBJECT(type));
 #endif
